### PR TITLE
Add SynthID (watermerking by Google DeepMind)

### PR DESCRIPTION
**Pull request by gante**
# What does this PR do?

Adds SynthID, a watermarking by DeepMind.

https://deepmind.google/technologies/synthid/

Applying watermarking and using a detector is added to `transfomers`. Training a detector is added as a research project.

### CODE REVIEW

**Review by ArthurZucker — COMMENTED** [review]
First path! Looks really good!

File: examples/research_projects/synthid_text/utils.py [hunk 1]
```diff
@@ -0,0 +1,427 @@
...
+        g_values = pad_to_len(g_values, max_length, True, 0, torch_device)
+
+        # We pad on left to get same max_length shape.
+        # g_values of shape [num_filtered_entries, max_length, depth].
+        all_masks.append(combined_mask)
+        all_g_values.append(g_values)
+    return all_masks, all_g_values
+
+
+def tpr_at_fpr(detector, detector_inputs, w_true, minibatch_size, target_fpr=0.01) -> torch.Tensor:
```
[the remark this answers is not in the document]
Comment (gante) [reply examples/research_projects/synthid_text/utils.py:168]: expanded docstring 👍

File: src/transformers/generation/watermarking.py [hunk 2]
```diff
@@ -237,3 +240,298 @@ def __call__(
...
+
+    This takes in g-values and returns p(g_values|watermarked).
+    """
+
+    def __init__(self, watermarking_depth: int):
+        """Initializes the model parameters."""
+        super().__init__()
+        self.watermarking_depth = watermarking_depth
+        self.beta = torch.nn.Parameter(-2.5 + 0.001 * torch.randn(1, 1, watermarking_depth))
+        self.delta = torch.nn.Parameter(0.001 * torch.randn(1, 1, self.watermarking_depth, watermarking_depth))
```
[the remark this answers is not in the document]
Comment (gante) [reply src/transformers/generation/watermarking.py:301]: It's not common, but we do have this pattern in other places ([e.g.](https://github.com/huggingface/transformers/blob/26370a4f18a44af529a84dd4c7a7fe7c37e67ed5/src/transformers/models/mllama/modeling_mllama.py#L928))

(I also have no idea how to set this specific initialization in `_init_weights ` 😅 )

File: src/transformers/generation/watermarking.py [hunk 3]
```diff
@@ -237,3 +240,298 @@ def __call__(
...
+
+        # [batch_size, seq_len, watermarking_depth, watermarking_depth]
+        x = torch.repeat_interleave(torch.unsqueeze(g_values, dim=-2), self.watermarking_depth, axis=-2)
+
+        # mask all elements above -1 diagonal for autoregressive factorization
+        x = torch.tril(x, diagonal=-1)
+
+        # [batch_size, seq_len, watermarking_depth]
+        # Long tensor doesn't work with einsum, so we need to switch to the same dtype as self.delta (FP32)
+        logits = torch.einsum("ijkl,ijkl->ijk", self.delta, x.type(self.delta.dtype)) + self.beta
```
[the remark this answers is not in the document]
Comment (gante) [reply src/transformers/generation/watermarking.py:329]: Good idea!

(the correct form is then `(self.delta[.., None,:] @ x[..., None]).squeeze()`)

**Review by ArthurZucker — APPROVED** [review]
Thanks for updating! 🤗

**Pull request merged**